### PR TITLE
Add projectile delay before despawn

### LINKED ISSUE #1662: Make projectile and floating toxin particles wait for a bit before being freed
To make projectiles leave a trail after it has hit something or despawns, so it doesn't disappear instantly. And the same for free floating toxins. A way to do it might be by turning off the emitter on despawn and have a timer that QueueFrees the particles on timeout

### PR DESCRIPTION
**Pull request by Apsum**
closes #1662 Make projectile and floating toxin particles wait for a bit before being freed #1662

### CODE REVIEW

File: src/microbe_stage/AgentProjectile.cs [hunk 1]
```diff
@@ -4,14 +4,18 @@
...
 public class AgentProjectile : RigidBody, ITimedLife
 {
     public float TimeToLiveRemaining { get; set; }
     public float Amount { get; set; }
     public AgentProperties Properties { get; set; }
     public Node Emitter { get; set; }
 
+    public Timer DespawnTimer;
+    //Delay has to be defined
+    public float Delaydespawn = 250;
```
Comment (Kasterisk) [inline src/microbe_stage/AgentProjectile.cs:16]: I think it would be better for this to be made as a constant in Constants.cs.

File: src/microbe_stage/AgentProjectile.cs [hunk 2]
```diff
@@ -4,14 +4,18 @@
...
 [JSONAlwaysDynamicType]
-[SceneLoadedClass("res://src/microbe_stage/AgentProjectile.tscn", UsesEarlyResolve = false)]
 public class AgentProjectile : RigidBody, ITimedLife
 {
     public float TimeToLiveRemaining { get; set; }
     public float Amount { get; set; }
     public AgentProperties Properties { get; set; }
     public Node Emitter { get; set; }
 
+    public Timer DespawnTimer;
```
Comment (Kasterisk) [inline src/microbe_stage/AgentProjectile.cs:14]: Should be a private field, so this wouldn't get assigned somewhere else.

File: src/microbe_stage/AgentProjectile.cs [hunk 3]
```diff
@@ -35,12 +46,30 @@ public void OnBodyEntered(Node body)
             }
         }
 
+
+        DespawnTimer.Start();
```
Comment (Kasterisk) [inline src/microbe_stage/AgentProjectile.cs:50]: The particle emitter in the AgentProjectile should also be turned off here for the residual trailing effect. Also there's an extra whitespace on top of this call, should be removed.

File: src/microbe_stage/AgentProjectile.cs [hunk 4]
```diff
@@ -4,14 +4,18 @@
...
     public float TimeToLiveRemaining { get; set; }
     public float Amount { get; set; }
     public AgentProperties Properties { get; set; }
     public Node Emitter { get; set; }
 
+    public Timer DespawnTimer;
+    //Delay has to be defined
+    public float Delaydespawn = 250;
+
+
```
Comment (Kasterisk) [inline src/microbe_stage/AgentProjectile.cs:18]: Extra whitespace.

**Review by Kasterisk — CHANGES_REQUESTED** [review]

File: src/microbe_stage/AgentProjectile.cs [hunk 5]
```diff
@@ -4,14 +4,18 @@
 ///   This is a shot agent projectile, does damage on hitting a cell of different species
 /// </summary>
 [JSONAlwaysDynamicType]
-[SceneLoadedClass("res://src/microbe_stage/AgentProjectile.tscn", UsesEarlyResolve = false)]
```
Comment (hhyyrylainen) [inline src/microbe_stage/AgentProjectile.cs:7]: Why did you remove this line? I'm pretty close to certain that this will break loading saves.

File: src/microbe_stage/AgentProjectile.cs [hunk 6]
```diff
@@ -21,6 +25,13 @@ public override void _Ready()
     {
         AddCollisionExceptionWith(Emitter);
         Connect("body_entered", this, "OnBodyEntered");
+
+        // Timer that delay despawn of projectiles
+        DespawnTimer = new Timer();
+        DespawnTimer.OneShot = true;
+        DespawnTimer.WaitTime = Delaydespawn;
+        DespawnTimer.Connect("timeout", this, "OnTimerTimeout");
+        AddChild(DespawnTimer);
```
Comment (hhyyrylainen) [inline src/microbe_stage/AgentProjectile.cs:34]: In order to support saving, this is not a very good approach. Instead a few new properties should be added like:
```
[JSONProperty]
bool FadingOut;
[JSONProperty]
float FadeTimeRemaining;
bool PhysicsDestroyed;
```

Then in _Process the time should be reduced, this destroyed when it reaches zero. And finally if PhysicsDestroyed is not true, then physics should be disabled and the velocity of this killed. That way saving should work without any complications.

File: src/microbe_stage/AgentProjectile.cs [hunk 7]
```diff
@@ -35,12 +46,30 @@ public void OnBodyEntered(Node body)
             }
         }
 
+
+        DespawnTimer.Start();
+    }
+
+    public void ApplyPropertiesFromSave(AgentProjectile projectile)
```
Comment (hhyyrylainen) [inline src/microbe_stage/AgentProjectile.cs:53]: This is the approach used by the old version of the save system. I'm quite sure that this method (at least with current master branch version), would never get called at all...

File: src/microbe_stage/AgentProjectile.cs [hunk 8]
```diff
@@ -35,12 +46,30 @@ public void OnBodyEntered(Node body)
...
+    public void OnTimerTimeout()
+    {
         Destroy();
     }
 
     private void Destroy()
     {
         // We should probably get some *POP* effect here.
-        this.DetachAndQueueFree();
+        QueueFree();
```
Comment (hhyyrylainen) [inline src/microbe_stage/AgentProjectile.cs:73]: Is there a reason why you changed this?

**Review by hhyyrylainen — CHANGES_REQUESTED** [review]
